### DISCUSSION TITLE
ffi issues

### DISCUSSION THREAD

**Discussion by sakulstra** [foundry-rs/foundry#4255] [Q&A]
Hello,
I wrote a foundry script with essentially logs some storage to a file, before and after the execution of some commands.
Afterwards i want to diff these two files, which works just fine via Makefile:

```Makefile
git-diff :
	@printf '%s\n%s\n%s\n' "\`\`\`diff" "$$(git diff --no-index --diff-algorithm=patience --ignore-space-at-eol ${before} ${after})" "\`\`\`" > diffs/${out}.md
```

Now i noticed that doing this, after the script execution is a bit cumbersome and i'd rather have the diffing be part of the foundry script, but I simply can't manage to.

I tried to essentially execute the whole command via `ffi` which threw a gazillion of rather weird errors(https://book.getfoundry.sh/cheatcodes/ffi?highlight=ffi#ffi).

So I tried reducing complexity to sth like `printf hello > test.txt` just to see if it works in general:
```
function testFfi() public {
    string[] memory inputs = new string[](4);
    inputs[0] = 'printf';
    inputs[1] = 'hello';
    inputs[2] = '>';
    inputs[3] = 'test.txt';

    vm.ffi(inputs);
  }
```

Which yields `2023-02-02T16:56:28.504122Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="printf: warning: ignoring excess arguments, starting with ‘>’\n"`.

Is there any way to do what i want to do via foundry?
Essentially I want to exec a command without caring for the response, so perhaps ffi might not be the correct tool for the job.

**Comment by DaniPopes** [comment]
The `> text.txt` is not part of the call to printf, it's a sh builtin: you're executing the command `printf` with 3 arguments, which fails. Same goes for expansions: `$(...)`
You would have to wrap the `printf` with a `sh` / `bash` command like so:
`sh -c "printf hello > test.txt"`:
```solidity
string[] memory inputs = new string[](4);
inputs[0] = "sh";
inputs[1] = "-c";
inputs[2] = "printf hello > test.txt";
vm.ffi(inputs);
```
